modules/katamari/src/main/ts/ephox/katamari/api/LazyValues.ts

```diff
@@ -1,7 +1,20 @@
 import { LazyValue } from './LazyValue';
 import * as AsyncValues from '../async/AsyncValues';
+import { clearTimeout, setTimeout } from '@ephox/dom-globals';
 
 /** par :: [LazyValue a] -> LazyValue [a] */
 export const par = function <T> (lazyValues: LazyValue<T>[]) {
   return AsyncValues.par(lazyValues, LazyValue.nu);
-};
\ No newline at end of file
+};
+
+export const withTimeout: <T>(baseFn: (completer: (value: T) => void) => void, ifTimeout: () => T, timeout: number) => LazyValue<T> = <T> (baseFn: (completer: (value: T) => void) => void, ifTimeout: () => T, timeout: number): LazyValue<T> =>
```

Comment:
I'm wondering about the architecture of this as it feels weird, to be honest. Would it be better if instead, `get` had 2 arguments: `callback` and `failure` (similar to fold). That way you wouldn't need a new method for this and instead could make `timeout` optional on the current implementation. eg:

```
const nu = function <T> (baseFn: (completer: (value: T) => void) => void, timeout?: number): LazyValue<T> {
```

Reply:
I didn't want to bake the concepts of failure or timeout into LazyValue, and force all callers to deal with that. It seems like a separate concern.

I usually prefer to add behaviours in layers. 

Our LazyValue here is a `Future`, but with slightly different semantics. A `Future` may return a value some time in the future. If you want a notion of failure, you make a `Future<Result>`. If you want a timeout, you create a `Future<Result>` that returns a failure result some time in the future. Maybe one caller might want a `Future<Option>`. Maybe another caller might want a `Future<Future<string>>`. 

When you bake these notions into the data structures, they become special purpose and stop being building blocks.

Reply:
Fair, though that only resolves it being a separate function. I still don't like the idea of the `ifTimeout` callback being passed in, as that goes against however everything else is done. If we take your example of `Future`/`Result`, then I like the idea of how `FutureResult` works.

Reply:
If we're wanting to avoid methods, then that becomes:
`LazyValue.withTimeout(LazyValue.nu(cb), timeout, errorFn).get()`

Reply:
We're not doing that yet though, as if you were to do that entirely without instance methods, then it'd be:
```
get(withTimeout(nu(cb), timeout, errorFn), getFn);
```
so it'd be consistent with that at least, in that all functions take a `LazyValue` as the first argument.

Reply:
Anyways, my original issue is that I wanted to try and keep the error callback and success callback closer together. Reason being having the error callback in the "constructor" is rather weird/code smells.

Reply:
True. Yes, we're not going quite that far, yet. 

Would you be happy with this:

`LazyValue.withTimeout(LazyValue.nu(cb), timeout, errorFn).get()`

or should we look at somethinge else?

Reply:
That's fine by me, thanks Dylan :+1:

Reply:
Cheers. Thanks for the feedback.

Reply:
Actually, there's a semantic problem with that. `LazyValue.nu` kicks off the computation immediately - before returning. If the base computation is long and synchronous, then the timer won't even be started until the base computation is complete.

Reply:
Ahh right... then let's just stick with the original I guess.